### PR TITLE
Rejection sink

### PR DESCRIPTION
**Pull request by rvesse**
Adds a new `RejectSink` which is basically just a `FilterSink` **but** it rejects with a `SinkException` rather than just dropping items.  Refactors how `FilterSink` is built to make it reusable for `RejectSink`.

Also cleaned up the test suite for basic sinks to split tests for each sink type out into their own test classes, plus renamed and refactored tests into Given, When, Then style.

Again this PR is consolidating pieces of work done elsewhere into a portable and reusable form for our Core Libraries.

### CODE REVIEW

**Review by robwtelicent — COMMENTED** [review]
I think we need to add the following dependencies to `pom.xml` in `projects-core`:
```
<dependency>
    <groupId>org.apache.commons</groupId>
    <artifactId>commons-lang3</artifactId>
</dependency>

<dependency>
    <groupId>io.opentelemetry</groupId>
    <artifactId>opentelemetry-api</artifactId>
</dependency>
```

File: projectors-core/src/main/java/io/telicent/smart/cache/projectors/sinks/FilterSink.java [hunk 1]
```diff
@@ -25,14 +25,20 @@
 import io.telicent.smart.cache.projectors.Library;
 import io.telicent.smart.cache.projectors.Sink;
 import io.telicent.smart.cache.projectors.SinkException;
+import io.telicent.smart.cache.projectors.sinks.builder.AbstractFilteringSinkBuilder;
 import io.telicent.smart.cache.projectors.sinks.builder.AbstractForwardingSinkBuilder;
 import org.apache.commons.lang3.StringUtils;
 
 import java.util.UUID;
+import java.util.function.Function;
```
Comment (robwtelicent) [inline projectors-core/src/main/java/io/telicent/smart/cache/projectors/sinks/FilterSink.java:33]: Appears to be unused

**Comment by rvesse** [comment]
> I think we need to add the following dependencies to `pom.xml` in `projectors-core`:
> 
> ```
> <dependency>
>     <groupId>org.apache.commons</groupId>
>     <artifactId>commons-lang3</artifactId>
> </dependency>
> 
> <dependency>
>     <groupId>io.opentelemetry</groupId>
>     <artifactId>opentelemetry-api</artifactId>
> </dependency>
> ```
> 
> Otherwise I see a lot of red in IntelliJ in `FilterSink.java`

Weird those should already be there, `mvn dependency:tree` output with `-Dincludes` for the relevant libraries:

> [INFO] --- dependency:3.8.1:tree (default-cli) @ projectors-core ---
[INFO] io.telicent.smart-caches:projectors-core:jar:0.25.0-SNAPSHOT
[INFO] +- io.telicent.smart-caches:observability-core:jar:0.25.0-SNAPSHOT:compile
[INFO] |  \- org.apache.commons:commons-lang3:jar:3.17.0:compile
[INFO] +- org.apache.jena:jena-arq:jar:5.2.0:compile
[INFO] |  \- org.apache.jena:jena-core:jar:5.2.0:compile
[INFO] |     \- org.apache.jena:jena-base:jar:5.2.0:compile
[INFO] |        +- org.apache.commons:commons-csv:jar:1.12.0:compile
[INFO] |        \- org.apache.commons:commons-compress:jar:1.27.1:compile
[INFO] \- org.apache.commons:commons-collections4:jar:4.4:compile

> [INFO] --- dependency:3.8.1:tree (default-cli) @ projectors-core ---
[INFO] io.telicent.smart-caches:projectors-core:jar:0.25.0-SNAPSHOT
[INFO] +- io.telicent.smart-caches:observability-core:jar:0.25.0-SNAPSHOT:compile
[INFO] |  \- io.opentelemetry:opentelemetry-api:jar:1.44.1:compile
[INFO] |     \- io.opentelemetry:opentelemetry-context:jar:1.44.1:compile
[INFO] \- io.opentelemetry:opentelemetry-sdk:jar:1.44.1:test
[INFO]    +- io.opentelemetry:opentelemetry-sdk-common:jar:1.44.1:test
[INFO]    +- io.opentelemetry:opentelemetry-sdk-trace:jar:1.44.1:test
[INFO]    |  \- io.opentelemetry:opentelemetry-api-incubator:jar:1.44.1-alpha:test
[INFO]    +- io.opentelemetry:opentelemetry-sdk-metrics:jar:1.44.1:test
[INFO]    \- io.opentelemetry:opentelemetry-sdk-logs:jar:1.44.1:test

So guessing IntelliJ is being awkward about resolving dependencies for you?

File: projectors-core/src/main/java/io/telicent/smart/cache/projectors/sinks/FilterSink.java (hunk 1 above)
Comment (rvesse) [reply projectors-core/src/main/java/io/telicent/smart/cache/projectors/sinks/FilterSink.java]: Cleaned up this and some other unused imports

**Review by TelicentPaul — APPROVED** [review]

**Comment by robwtelicent** [comment]
> mvn dependency:tree

I think the dependencies are being pulled in transitively, so it complies OK, but InitelliJ is probably complaining because if you run `mvn dependency:analyze` it shows them as being undeclared.

**Comment by rvesse** [comment]
> > mvn dependency:tree
> 
> I think the dependencies are being pulled in transitively, so it complies OK, but InitelliJ is probably complaining because if you run `mvn dependency:analyze` it shows them as being undeclared.

Good point, generally we've somewhat relied on transitive dependencies so we don't end up redefining dependencies everywhere but may be worth reviewing that a bit to see if there's scope for minimising the dependencies in some modules e.g. I note `projectors-core` depends on `jena-arq` but never actually uses any Jena APIs

**Pull request merged**